Minor Fixes

ExtremeFiretop opened the pull request:
Minor Fixes

1. Added a check to validate the downloaded firmware version by MerlinAU is infact the version found by the router's webs_update.sh script.

Seems to me what happened at some point for my RT-AX92U router is @GNUton built the firmwares; (lets say at midnight) and his Github action automatically published it. Shortly after it was automatically published, the Webs-Update.sh script ran and found the update.

MerlinAU then ran a few hours later by cron job, by that point the builds were hidden. (Gnuton had unpublished or hidden the builds by 3AM) But because they were published and the webs_update.sh script detected it, and started to flash.

So instead MerlinAU downloaded the latest 388.8_2 and downgraded me from the latest alpha lol. 

![image](https://github.com/user-attachments/assets/3678bc51-3247-431b-800f-f329269dac62)
![image](https://github.com/user-attachments/assets/09cdadae-c101-4bf0-9a0b-21f8871f57d8)

This additional check will cancel the script if this happens again and the versions don't match. Along with additional Gnuton specific changes to make sure we grab the right version by parsing for matching version and matching Product ID.

2. Fixed the post email upgrade weirdness we found in January here: 
https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/395#issuecomment-2614229361

![image](https://github.com/user-attachments/assets/b6145d62-50b9-45eb-a031-bdd197490fff)

In short; we shouldn't be using: `_GetLatestFWUpdateVersionFromRouter_ 1` because that does not split on the dash. That was a me mistake when I implemented this I think?

While `_GetLatestFWUpdateVersionFromRouter_`  does split the version information at the dash for things like rog, or -gnuton.

ExtremeFiretop commented on MerlinAU.sh:
```diff
@@ -2846,7 +2846,7 @@ _CreateEMailContent_()
    if ! "$offlineUpdateTrigger"
    then
-        fwNewUpdateVersion="$(_GetLatestFWUpdateVersionFromRouter_ 1)"
+        fwNewUpdateVersion="$(_GetLatestFWUpdateVersionFromRouter_)"
```
This is the post upgrade email fix. Simple and easy

ExtremeFiretop commented on MerlinAU.sh:
```diff
@@ -4545,29 +4545,52 @@ _GetLatestFWUpdateVersionFromGithub_()
         search_type="pure\|squashfs\|ubi"
     fi
 
+    if ! "$offlineUpdateTrigger"
```
The rest below is to fix the issue of not validating the downloaded version is the correct version we got notified about from Webs_update.sh

ExtremeFiretop commented on MerlinAU.sh:
```diff
@@ -8384,6 +8433,16 @@ Please manually update to version ${GRNct}${MinSupportedFirmwareVers}${NOct} or
         fi
     fi
 
+    if ! "$offlineUpdateTrigger"
```
Finally; add a final check before flashing to validate it all and cancel if they still don't match somehow.
The above code is Gnuton specific, this catch is for all firmwares, Gnuton or Merlin

ExtremeFiretop commented on MerlinAU.sh:
```diff
@@ -4545,29 +4545,52 @@ _GetLatestFWUpdateVersionFromGithub_()
+            # Extract the version portion from the URL
+            local version="$(echo "$url_item" \
+                | grep -oE "${PRODUCT_ID}_[^ ]*\.(w|pkgtb)" \
+                | sed "s/${PRODUCT_ID}_//;s/.w$//;s/.pkgtb$//;s/.ubi$//;s/_/./g" | head -n1)"
```
With the head -n1 here, I was getting multiple lines with the existing code:

```
Admin@GT-BE98_Pro:/tmp/home/root# echo "$version"
4-gnuton1
3004.388.8.4-gnuton1.puresqubi
```

TheS1R replied:
> Without the head -n1 here, I was getting multiple lines with the existing code:
> 
> ```
> Admin@GT-BE98_Pro:/tmp/home/root# echo "$version"
> 4-gnuton1
> 3004.388.8.4-gnuton1.puresqubi
> ```

That darn gnuton is continuing to give you grief!

ExtremeFiretop replied:
> > Without the head -n1 here, I was getting multiple lines with the existing code:
> > ```
> > Admin@GT-BE98_Pro:/tmp/home/root# echo "$version"
> > 4-gnuton1
> > 3004.388.8.4-gnuton1.puresqubi
> > ```
> 
> That darn gnuton is continuing to give you grief!

We'll get there. 

It's a little bit different since it's dealing with GitHub instead of Sourceforge and they don't package them the same, and they have their own UIs (TUF instead of ROG) and so on,  it's always fun to identify spots to improve. 

Both these things were on my to do list for a while but I never got around to it, and no one reported anything hahaha

ExtremeFiretop replied:
This is probably one of the only add-ons you'll find where the logic flow takes a completely different route between Merlin and Gnuton, usually scripts just run on Gnuton with little to no changes.

But because we are dealing with flashing the firmware automatically that does have a different process between them haha

Martinski4GitHub commented:
> Minor Fixes
> 
>     1. Added a check to validate the downloaded firmware version by MerlinAU is infact the version found by the router's webs_update.sh script.
> 
> 
> Seems to me what happened at some point for my RT-AX92U router is @gnuton built the firmwares; (lets say at midnight) and his Github action automatically published it. Shortly after it was automatically published, the Webs-Update.sh script ran and found the update.
> 
> MerlinAU then ran a few hours later by cron job, by that point the builds were hidden. (Gnuton had unpublished or hidden the builds by 3AM) But because they were published and the webs_update.sh script detected it, and started to flash.
> 
> So instead MerlinAU downloaded the latest 388.8_2 and downgraded me from the latest alpha lol.
> 
> ...
> 
> This additional check will cancel the script if this happens again and the versions don't match. Along with additional Gnuton specific changes to make sure we grab the right version by parsing for matching version and matching Product ID.
> 
>     2. Fixed the post email upgrade weirdness we found in January here:
>        [New Checkmarks on WebGUI #395 (comment)](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/395#issuecomment-2614229361)
> 
> ....
> 
> In short; we shouldn't be using: `_GetLatestFWUpdateVersionFromRouter_ 1` because that does not split on the dash. That was a me mistake when I implemented this I think?
> 
> While `_GetLatestFWUpdateVersionFromRouter_` does split the version information at the dash for things like rog, or -gnuton.
>
>
Very good finds, bud!! I can only assume that this scenario with a Gnuton F/W build being published and then, hours later, being "recalled" (i.e. unpublished or hidden somehow) is an extremely rare event, and that's why we've never seen a user reporting that MerlinAU "downgraded" the installed F/W version. I don't recall reading about this same scenario happening with RMerlin's F/W releases at all.

In any case, it's always good to add code that double-checks & verifies that the correct F/W release version is in fact being installed. Excellent job, bud!

Martinski4GitHub reviewed (approved):
Good to go!!

ExtremeFiretop commented:
> > Minor Fixes
> > ```
> > 1. Added a check to validate the downloaded firmware version by MerlinAU is infact the version found by the router's webs_update.sh script.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Seems to me what happened at some point for my RT-AX92U router is @gnuton built the firmwares; (lets say at midnight) and his Github action automatically published it. Shortly after it was automatically published, the Webs-Update.sh script ran and found the update.
> > MerlinAU then ran a few hours later by cron job, by that point the builds were hidden. (Gnuton had unpublished or hidden the builds by 3AM) But because they were published and the webs_update.sh script detected it, and started to flash.
> > So instead MerlinAU downloaded the latest 388.8_2 and downgraded me from the latest alpha lol.
> > ...
> > This additional check will cancel the script if this happens again and the versions don't match. Along with additional Gnuton specific changes to make sure we grab the right version by parsing for matching version and matching Product ID.
> > ```
> > 2. Fixed the post email upgrade weirdness we found in January here:
> >    [New Checkmarks on WebGUI #395 (comment)](https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/395#issuecomment-2614229361)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ....
> > In short; we shouldn't be using: `_GetLatestFWUpdateVersionFromRouter_ 1` because that does not split on the dash. That was a me mistake when I implemented this I think?
> > While `_GetLatestFWUpdateVersionFromRouter_` does split the version information at the dash for things like rog, or -gnuton.
> 
> Very good finds, bud!! I can only assume that this scenario with a Gnuton F/W build being published and then, hours later, being "recalled" (i.e. unpublished or hidden somehow) is an extremely rare event, and that's why we've never seen a user reporting that MerlinAU "downgraded" the installed F/W version. I don't recall reading about this same scenario happening with RMerlin's F/W releases at all.
> 

Thank you! Yes i think it's fairly rare, but we should probably build in a "safety" against it. I won't lie, it really confused me to wake up to my router DOWNGRADED by MerlinAU hahaha

> In any case, it's always good to add code that double-checks & verifies that the correct F/W release version is in fact being installed. 

Agreed!

ExtremeFiretop commented:
Btw check SNB, spdMerlin reports. I'm troubleshooting as we speak
